lib/discovery.js

```diff
@@ -0,0 +1,198 @@
+import { TypedEmitter } from "tiny-typed-emitter";
+import Hyperswarm from "hyperswarm";
+
+import { LocalPeers } from "./localpeers.js";
+
+/**
+ * @typedef {Object} DiscoveryEvents
+ * @property {(connection, info) => void} peer
+ * @property {(error: Error) => void} error
+ */
+
+/**
+ * @extends {TypedEmitter<DiscoveryEvents>}
+ */
+export class Discovery extends TypedEmitter {
+  #peers = new Map();
+  #discover = new Map();
+  #dht;
+  #mdns;
+
+  /**
+   * @param {object} options
+   * @param {number} options.port
+   * @param {object} options.keyPair
+   * @param {Buffer} options.keyPair.publicKey
+   * @param {Buffer} options.keyPair.secretKey
+   * @param {boolean|object} [options.dht=true]
+   * @param {boolean|object} [options.mdns=true]
+   */
+  constructor({ keyPair, port, dht = true, mdns = true }) {
+    super();
+
+    if (dht !== false) {
+      this.dhtOptions = typeof dht === "object" ? dht : {};
+      this.dhtOptions.keyPair = keyPair;
+
+      this.#dht = new Hyperswarm(this.dhtOptions);
+      this.#dht.on("connection", this._onDhtPeer.bind(this));
+    }
+
+    if (mdns) {
+      const mdnsOptions = typeof mdns === "object" ? mdns : {};
+      mdnsOptions.keyPair = keyPair;
+      mdnsOptions.port = port;
+      mdnsOptions.name = "mapeo";
+
+      this.#mdns = new LocalPeers(mdnsOptions);
+      this.#mdns.on("connection", this._onMdnsPeer.bind(this));
+    }
+  }
+
+  get peers() {
+    return Array.from(this.#peers.values());
+  }
+
+  _onDhtPeer(connection, info) {
+    const publicKey = connection.remotePublicKey.toString("hex");
+    const peer = this.#peers.get(publicKey);
+
+    if (!peer) {
+      const peerInfo = new PeerInfo({
+        topics: info.topics,
+        host: connection.remoteHost,
+        port: connection.remotePort,
+        discoveryType: "dht",
+        identityPublicKey: publicKey,
+      });
+
+      this.#peers.set(publicKey, {
+        connection,
+        peerInfo,
+      });
+
+      this.emit("peer", connection, peerInfo);
+    } else {
+      this.#dht.leavePeer(publicKey);
```

Comment:
Do I understand this correctly: Is this to de-dupe connections made on both mdns and dht? Prioritizing mdns (if it is first)? Maybe we could do an additional check on the IP range so that we could also prioritize mdns if it is not first?